tests/unit/capi/GEOSDensifyTest.cpp

```diff
@@ -0,0 +1,214 @@
+//
+// Test Suite for C-API GEOSBuffer and GEOSBufferWithStyle
+
+#include <tut/tut.hpp>
+// geos
+#include <geos_c.h>
+
+#include "capi_test_utils.h"
+
+
+namespace tut {
+//
+// Test Group
+//
+
+// Common data used in test cases.
+struct test_capigeosdensify_data : public capitest::utility {
+    GEOSGeometry* geom1_;
+    GEOSGeometry* geom2_;
+    GEOSWKTWriter* w_;
+    char* wkt_;
+
+    test_capigeosdensify_data()
+        : geom1_(nullptr), geom2_(nullptr), w_(nullptr)
+    {
+        initGEOS(notice, notice);
+        w_ = GEOSWKTWriter_create();
+        GEOSWKTWriter_setTrim(w_, 1);
+    }
+
+    ~test_capigeosdensify_data()
+    {
+        GEOSGeom_destroy(geom1_);
+        GEOSGeom_destroy(geom2_);
+        GEOSWKTWriter_destroy(w_);
+        geom1_ = nullptr;
+        geom2_ = nullptr;
+        wkt_ = nullptr;
+        finishGEOS();
+    }
+
+};
+
+typedef test_group<test_capigeosdensify_data> group;
+typedef group::object object;
+
+group test_capigeosdensify_group("capi::GEOSDensify");
+
+//
+// Test Cases
+//
+
+// Densify with a tolerance slightly larger than length of all edges.
+// Result should match inputs.
+template<>
+template<>
+void object::test<1>()
+{
+    geom1_ = GEOSGeomFromWKT("POLYGON ((0 0, 10 0, 10 10, 0 10, 0 0), (1 1, 1 2, 2 2, 2 1, 1 1))");
+    ensure(geom1_ != nullptr);
+    GEOSSetSRID(geom1_, 3857);
+
+    geom2_ = GEOSDensify(geom1_, 10.0);
+
+    ensure("result not null", geom2_ != nullptr);
+    ensure_geometry_equals(geom2_, geom1_);
+    ensure_equals("result SRID == expected SRID", GEOSGetSRID(geom2_), 3857);
+}
+
+
+// Densify with a tolerance that evenly subdivides all outer and inner edges.
+template<>
+template<>
+void object::test<2>()
+{
+    geom1_ = GEOSGeomFromWKT("POLYGON ((0 0, 10 0, 10 10, 0 10, 0 0), (1 1, 1 7, 7 7, 7 1, 1 1))");
+    ensure(geom1_ != nullptr);
+    GEOSSetSRID(geom1_, 3857);
+
+    geom2_ = GEOSDensify(geom1_, 5.0);
+
+    ensure("result not null", geom2_ != nullptr);
+    ensure_geometry_equals(
+        geom2_,
+        "POLYGON ((0 0, 5 0, 10 0, 10 5, 10 10, 5 10, 0 10, 0 5, 0 0), (1 1, 1 4, 1 7, 4 7, 7 7, 7 4, 7 1, 4 1, 1 1))");
+
+    ensure_equals("result SRID == expected SRID", GEOSGetSRID(geom2_), 3857);
```

Comment:
Sure would be nice to reduce the amount of code for tests like this (not just here, for all GEOS unit tests).

Any ideas, @dbaston ?  Perhaps some more helper functions?

Reply:
If I were writing it I would drop lines 77 and 82, but overall it seems pretty reasonable to me.

Reply:
I don't have a problem with those lines per se (although if they are redundant would be good to remove them).  The real problem is how many lines of code are repeated in each test.  The only information content is the input WKT, the densify tolerance, and the result WKT. In JTS I factor out the boilerplate into a separate function which is called by each test.  I'm just not sure if the tut templated approach supports that?

Reply:
It turns out that factoring out the test common code into a support function is very doable.  See [`GEOSMinimumClearanceTest`](https://git.osgeo.org/gitea/geos/geos/src/branch/master/tests/unit/capi/GEOSMinimumClearanceTest.cpp).cpp for an example.

@brendan-ward how do you feel about making this improvement?

Reply:
Yes, some of the tests do something like `doDensifierTest`. I actually find those harder to work with, mostly because our test framework has no way to run an individual test. So if you want to debug a test and the test code itself is in a common function, you have to comment out all of the remaining tests.

Reply:
Another negative is when trying to read a test, you have to follow a trail of function calls to figure out what `doDensifierTest` even means. I think the style of good test code is a bit different from implementation code, and good test code is about easy readability, not compactness.

Reply:
TIL the software engineering consultants call this "DAMP vs DRY"...there's always an acronym (or two)!

Reply:
>  So if you want to debug a test and the test code itself is in a common function, you have to comment out all of the remaining tests.

I don't see how this is affected by factoring out the common code?

Reply:
> Another negative is when trying to read a test, you have to follow a trail of function calls to figure out what `doDensifierTest` even means. 

Usually there is only a single function that needs to be understood.  And you could make the opposite argument - by using a common function you only need to understand that function to know what every test that uses it does, whereas if the code is duplicated you can't assume that tests always do the same thing, so need to read them individually (and keep them all in mind looking for minor differences).  These seems way more error-prone - there is a serious risk of some part of the test being omitted in a single test, which is unlikely to be noticed.

Reply:
Yeah, it's a balance. I've found having support functions tucked away in another file makes things more painful, but having them in the actual Test file is a-OK. Also, we can run individual *tests* (not assertions) with `./bin/test_geos_unit geos::io::WKBReader 6` so having more numbered chunks is often preferable to a long list of assertions, or a loop over a data array (my nemesis).

Reply:
> Also, we can run individual tests (not assertions) with ./bin/test_geos_unit geos::io::WKBReader 6

How am I only learning this now? 🤦

Reply:
I only found it out a couple months ago! I was like... "what if I add the number..."

Reply:
(Also, the numbers don't have to be sequential)

Reply:
> And you could make the opposite argument

Of course. My experience makes me less inclined to write something today the way I wrote the minimum clearance tests 5 years ago. Your experience seems to pull you in the other direction. That's fine.